src/Orleans/Serialization/TypeUtilities.cs

```diff
@@ -20,54 +21,63 @@ internal static bool IsOrleansPrimitive(this Type t)
             return typeInfo.IsPrimitive || typeInfo.IsEnum || t == typeof(string) || t == typeof(DateTime) || t == typeof(Decimal) || (typeInfo.IsArray && typeInfo.GetElementType().IsOrleansPrimitive());
         }
 
-        static readonly Dictionary<RuntimeTypeHandle, bool> shallowCopyableValueTypes = new Dictionary<RuntimeTypeHandle, bool>();
+        static readonly ConcurrentDictionary<Type, bool> shallowCopyableTypes = new ConcurrentDictionary<Type, bool>();
         static readonly Dictionary<RuntimeTypeHandle, string> typeNameCache = new Dictionary<RuntimeTypeHandle, string>();
         static readonly Dictionary<RuntimeTypeHandle, string> typeKeyStringCache = new Dictionary<RuntimeTypeHandle, string>();
         static readonly Dictionary<RuntimeTypeHandle, byte[]> typeKeyCache = new Dictionary<RuntimeTypeHandle, byte[]>();
 
         static TypeUtilities()
         {
-            shallowCopyableValueTypes[typeof(Decimal).TypeHandle] = true;
```

Comment:
Sorry, I don't understand this comment. Do you mean that using `typeof()` instead of `typeof().TypeHandle` as a key here is more efficient? I don't understand why.